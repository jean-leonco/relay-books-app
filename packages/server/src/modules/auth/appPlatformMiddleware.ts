import Application from 'koa';

import { GraphQLContext } from '../../types';
import { getPlatform, PLATFORM } from '../../common/utils';

const appPlatformMiddleware: Application.Middleware<any, GraphQLContext> = async (ctx, next) => {
  if (ctx.request.url === '/healthz') {
    return await next();
  }

  const { appplatform } = ctx.header;
  const platform = getPlatform(appplatform);

  if (platform === PLATFORM.UNKNOWN) {
    ctx.status = 403;
    ctx.body = {
      data: null,
      errors: [{ message: 'Forbidden', severity: 'WARNING' }],
    };

    return;
  }
  return await next();
};

export default appPlatformMiddleware;
