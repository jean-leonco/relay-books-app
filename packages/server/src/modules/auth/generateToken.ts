import jwt from 'jsonwebtoken';
import { addMinutes, addSeconds } from 'date-fns';

import { SESSION_TOKEN_SCOPES, IUser, SessionToken, ISessionToken } from '../../models';
import { GraphQLContext } from '../../types';

import { JWT_KEY } from '../../common/config';

const twelveHoursInMinutes = 60 * 12;

export const getExpirationTimeByScope = (scope: string): number | null => {
  if (Object.values(SESSION_TOKEN_SCOPES).indexOf(scope) === -1) {
    return null;
  }

  switch (scope) {
    case SESSION_TOKEN_SCOPES.SESSION:
      return null;
    case SESSION_TOKEN_SCOPES.RESET_PASSWORD:
      return twelveHoursInMinutes;
    default:
      return null;
  }
};

export const jwtSign = (sessionToken: ISessionToken, secondsToExpire?: number): string => {
  const expiresIn = secondsToExpire || getExpirationTimeByScope(sessionToken.scope);

  return jwt.sign(
    {
      id: sessionToken.user ? sessionToken.user : sessionToken.ip,
      scope: sessionToken.scope,
      sessionToken: sessionToken._id,
    },
    JWT_KEY,
    expiresIn ? { expiresIn: expiresIn * 60 } : {},
  );
};

export const generateToken = async (
  ctx: GraphQLContext,
  user: IUser | null,
  channel: string,
  scope: string,
  expiresIn?: number, // in seconds
): Promise<string | null> => {
  const { koaContext } = ctx;

  const { ip } = koaContext.request;
  const scopeExpirationTime = getExpirationTimeByScope(scope); // in minutes
  const scopeExpiresIn = scopeExpirationTime ? addMinutes(new Date(), scopeExpirationTime) : null;

  let sessionToken;

  // only reuse tokens that has user and the scope is SESSION
  if (user && scope === SESSION_TOKEN_SCOPES.SESSION) {
    sessionToken = await SessionToken.findOne({
      user: user._id,
      ip,
      channel,
      scope,
      isActive: true,
      isBlocked: false,
    }).lean<ISessionToken>();
  }

  if (sessionToken) {
    sessionToken = await SessionToken.findOneAndUpdate(
      { _id: sessionToken._id },
      { expiresIn: expiresIn ? addSeconds(new Date(), expiresIn) : scopeExpiresIn },
    ).lean<ISessionToken>();
  } else {
    sessionToken = await new SessionToken({
      ...(user ? { user: user._id } : {}),
      ip,
      channel,
      scope,
      expiresIn: expiresIn ? addSeconds(new Date(), expiresIn) : scopeExpiresIn,
    }).save();
  }

  return jwtSign(sessionToken!, expiresIn);
};

export const hasReachedMaximumNumberOfSessions = async (user: IUser, scope: string, numberOfSessions: number) => {
  const sessions = await SessionToken.find({
    user: user._id,
    scope,
    isActive: true,
    expiresIn: { $gte: new Date() },
  }).lean();

  return sessions.length >= numberOfSessions;
};
