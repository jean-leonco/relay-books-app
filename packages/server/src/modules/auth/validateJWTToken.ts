import jwt, { TokenExpiredError } from 'jsonwebtoken';

import { GraphQLContext } from '../../types';

import { SessionToken, ISessionToken } from '../../models';

import { JWT_KEY } from '../../common/config';
import { t } from '../../locales/helpers';

export interface DecodedToken {
  id?: string;
  scope?: string;
  sessionToken?: string;
  expired?: boolean;
  iat?: number;
}

interface ValidateTokenResult {
  token: DecodedToken | null;
  expired?: boolean;
  error?: string;
  detailedError?: string;
  abortRequest?: boolean;
  status?: number;
}

const validateJWTToken = async (context: GraphQLContext, token: string): Promise<ValidateTokenResult> => {
  try {
    const decodedToken: DecodedToken | string = jwt.verify(token, JWT_KEY);

    if (decodedToken == null) {
      return {
        token: null,
        error: t('auth', 'TokenInvalid'),
        detailedError: t('auth', 'TokenInvalid'),
        status: 403,
        abortRequest: true,
      };
    }

    if (typeof decodedToken === 'string') {
      return {
        token: null,
        error: t('auth', 'TokenInvalid'),
        detailedError: t('auth', 'TokenInvalid'),
        status: 403,
        abortRequest: true,
      };
    }

    if (!decodedToken.id || !decodedToken.scope || !decodedToken.sessionToken) {
      return {
        token: null,
        error: t('auth', 'TokenInvalid'),
        detailedError: t('auth', 'TokenInvalidSession'),
        status: 403,
        abortRequest: true,
      };
    }

    const sessionToken = await SessionToken.findOne({
      _id: decodedToken.sessionToken,
      user: decodedToken.id,
      scope: decodedToken.scope,
      isActive: true,
      isBlocked: false,
    }).lean<ISessionToken>();

    if (!sessionToken) {
      return {
        token: null,
        error: t('auth', 'InvalidSession'),
        detailedError: t('auth', 'InvalidSessionDetailed'),
        status: 401,
        abortRequest: true,
      };
    }

    if (sessionToken.scope !== decodedToken.scope) {
      return {
        token: null,
        error: t('auth', 'TokenInvalid'),
        detailedError: t('auth', 'InvalidSessionScope'),
        status: 403,
        abortRequest: true,
      };
    }

    return { token: decodedToken };
  } catch (err) {
    if (err instanceof TokenExpiredError) {
      return {
        token: null,
        expired: true,
        error: t('auth', 'TokenExpired'),
        detailedError: t('auth', 'TokenExpiredDetailed', { expired: err.expiredAt }),
        status: 401,
        abortRequest: true,
      };
    }

    if (err.message === 'invalid signature') {
      return {
        token: null,
        error: t('auth', 'TokenInvalid'),
        detailedError: t('auth', 'TokenInvalidSignature'),
        status: 403,
        abortRequest: true,
      };
    }

    if (err.name === 'JsonWebTokenError' || err.message === 'invalid token') {
      return {
        token: null,
        error: t('auth', 'TokenInvalid'),
        detailedError: t('auth', 'TokenInvalidSignature'),
        status: 403,
        abortRequest: true,
      };
    }

    return {
      token: null,
      error: t('auth', 'TokenInvalid'),
      detailedError: t('auth', 'TokenUnknownError'),
      status: 403,
      abortRequest: true,
    };
  }
};

export default validateJWTToken;
