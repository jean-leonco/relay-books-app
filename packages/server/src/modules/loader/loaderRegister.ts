import { GraphQLDataloaders } from '../../types';

const loaders: {
  [Name in keyof GraphQLDataloaders]: () => GraphQLDataloaders[Name];
} = {} as any;

const registerLoader = <Name extends keyof GraphQLDataloaders>(
  key: Name,
  getLoader: () => GraphQLDataloaders[Name],
) => {
  loaders[key] = getLoader as any;
};

const getDataloaders = (): GraphQLDataloaders =>
  (Object.keys(loaders) as (keyof GraphQLDataloaders)[]).reduce(
    (prev, loaderKey) => ({
      ...prev,
      [loaderKey]: loaders[loaderKey](),
    }),
    {},
  ) as any;

export { registerLoader, getDataloaders };
