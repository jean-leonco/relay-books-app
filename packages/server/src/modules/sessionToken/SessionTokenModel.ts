import mongoose, { Document, Model, Types } from 'mongoose';

import { PLATFORM } from '../../common/utils';
import { isActiveMongooseField } from '../../core/mongoose/withMongooseFields';

const { ObjectId } = mongoose.Schema.Types;

export enum SESSION_TOKEN_SCOPES {
  RESET_PASSWORD = 'BOOKSAPP:RESET_PASSWORD', // used when user requested a password reset
  SESSION = 'BOOKSAPP:SESSION', // token for a valid user session
}

const Schema = new mongoose.Schema(
  {
    user: {
      type: ObjectId,
      ref: 'User',
      description: 'User owner of this token.',
    },
    ip: {
      type: String,
      description: 'Agent IP that came from request headers.',
      required: true,
      index: true,
    },
    channel: {
      type: String,
      required: true,
      default: PLATFORM.APP,
      description: 'Channel that this token belongs too. ex: APP',
      enum: Object.values(PLATFORM),
    },
    name: {
      type: String,
      description: 'Name of this token. ex: Jean Linux',
    },
    scope: {
      type: String,
      description: 'Scope of this session token. ex: SESSION',
    },
    expiresIn: {
      type: Date,
      description: 'When this session token expires.',
    },
    isBlocked: {
      type: Boolean,
      description: 'If the token is blocked.',
      default: false,
      required: true,
      index: true,
    },
    ...isActiveMongooseField,
  },
  {
    timestamps: {
      createdAt: 'createdAt',
      updatedAt: 'updatedAt',
    },
    collection: 'SessionToken',
  },
);

export interface ISessionToken extends Document {
  user: Types.ObjectId;
  ip: string;
  channel: string;
  isActive: boolean;
  isBlocked: boolean;
  name: string;
  scope: string;
  expiresIn: Date | null;
  createdAt: Date;
  updatedAt: Date;
}

const SessionTokenModel: Model<ISessionToken> = mongoose.model('SessionToken', Schema);

export default SessionTokenModel;
