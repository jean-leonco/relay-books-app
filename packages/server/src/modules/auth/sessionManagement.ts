import { ParameterizedContext } from 'koa';

import { GraphQLContext } from '../../types';
import { SESSION_TOKEN_SCOPES } from '../../models';
import { User, UserLoader } from '../../loader';

import { t } from '../../locales/helpers';

import validateJWTToken from './validateJWTToken';

interface AuthorizationHeader {
  token: null | string;
  error: string | null;
  detailedError: string | null;
}

export interface ContextFromTokenResult {
  user: User | null;
  error?: string | null;
  detailedError?: string | null;
  abortRequest: boolean;
  unauthorized: boolean;
  status?: number;
}

export function parseAuthorizationHeader(token: string | null): AuthorizationHeader {
  if (!token || token === 'null' || token === 'undefined') {
    return { token: null, error: null, detailedError: null };
  }

  if (!token.match(/JWT /)) {
    return {
      token: null,
      error: t('auth', 'TokenInvalid'),
      detailedError: t('auth', 'TokenAuthHeaderMissingJwt'),
    };
  }

  const cleanToken = token.replace(/^JWT /, '').trim();

  if (cleanToken === '[object Object]') {
    return {
      token: null,
      error: t('auth', 'TokenInvalid'),
      detailedError: t('auth', 'TokenAuthHeader'),
    };
  }

  return { token: cleanToken, error: null, detailedError: null };
}

export const getUser = async (
  ctx: ParameterizedContext & GraphQLContext,
  authorization: string | null,
): Promise<ContextFromTokenResult> => {
  const defaultReturn = { user: null, abortRequest: false, unauthorized: false };

  const defaultInvalidToken = { user: null, abortRequest: true, unauthorized: true };

  if (!authorization) {
    return defaultReturn;
  }

  const decodedToken = await validateJWTToken(ctx, authorization);
  if (!decodedToken.token || decodedToken.error) {
    return {
      ...defaultInvalidToken,
      error: decodedToken.error,
      detailedError: decodedToken.detailedError,
      status: decodedToken.status,
    };
  }

  const { token } = decodedToken;

  const validSessions = Object.values(SESSION_TOKEN_SCOPES);

  // do not try to create session if it isn't a session token
  if (!token.scope || !validSessions.includes(token.scope)) {
    return { ...defaultReturn, unauthorized: true };
  }

  if (token.scope !== SESSION_TOKEN_SCOPES.SESSION) {
    return {
      ...defaultInvalidToken,
      error: t('auth', 'TokenInvalid'),
      detailedError: t('auth', 'InvalidSessionScope'),
      status: 401,
    };
  }

  try {
    const user = await UserLoader.load(ctx, token.id, true);

    if (!user) {
      return {
        ...defaultInvalidToken,
        error: t('auth', 'TokenInvalid'),
        detailedError: t('auth', 'UserNotFound'),
        status: 403,
      };
    }

    if (!user.isActive || user.removedAt) {
      return {
        ...defaultInvalidToken,
        error: t('auth', 'TokenInvalid'),
        detailedError: !user.isActive ? t('auth', 'UserDeactivated') : t('auth', 'UserNotFound'),
        status: 401,
      };
    }

    return { ...defaultReturn, user };
  } catch (err) {
    return defaultInvalidToken;
  }
};

export async function auth(ctx: ParameterizedContext & GraphQLContext): Promise<ContextFromTokenResult> {
  const { authorization } = ctx.header;

  const authHeader = parseAuthorizationHeader(authorization);

  if (authHeader.error) {
    return {
      user: null,
      error: authHeader.error,
      detailedError: authHeader.detailedError,
      abortRequest: true,
      unauthorized: true,
    };
  }

  return await getUser(ctx, authHeader.token);
}
