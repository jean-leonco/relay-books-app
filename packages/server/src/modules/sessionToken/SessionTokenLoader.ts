import DataLoader from 'dataloader';
import { Types } from 'mongoose';
import { isPast } from 'date-fns';

import { connectionFromMongoCursor, mongooseLoader } from '@entria/graphql-mongoose-loader';

import { GraphQLContext } from '../../types';

import { IUser } from '../user/UserModel';

import SessionTokenModel, { ISessionToken } from './SessionTokenModel';

export default class SessionToken {
  public registeredType = 'SessionToken';

  id: string;
  _id: Types.ObjectId;
  user: Types.ObjectId;
  ip: string;
  channel: string;
  scope: string;
  isActive: boolean;
  isBlocked: boolean;
  name: string;
  createdAt: Date;
  updatedAt: Date;
  expiresIn: Date | null;

  constructor(data: ISessionToken) {
    this.id = data.id || data._id;
    this._id = data._id;
    this.user = data.user;
    this.ip = data.ip;
    this.channel = data.channel;
    this.scope = data.scope;
    this.isActive = data.isActive;
    this.isBlocked = data.isBlocked;
    this.name = data.name;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
    this.expiresIn = data.expiresIn;
  }
}

export const getLoader = () => new DataLoader(ids => mongooseLoader(SessionTokenModel, ids));

// @TODO - improve this
const viewerCanSee = () => true;

export const load = async (context: GraphQLContext, id: string) => {
  if (!id) {
    return null;
  }

  try {
    const data = await context.dataloaders.SessionTokenLoader.load(id.toString());

    if (!data) {
      return null;
    }

    return viewerCanSee() ? new SessionToken(data) : null;
  } catch (err) {
    return null;
  }
};

export const clearCache = ({ dataloaders }: GraphQLContext, id: string) =>
  dataloaders.SessionTokenLoader.clear(id.toString());

// @TODO - do not query with isBlocked so we can validate if exists a session token
// blocked for this user and abort token generation process
export const loadValidSessionToken = async (
  ctx: GraphQLContext,
  user: IUser,
  channel: string,
  scope: string,
): Promise<ISessionToken | null> => {
  const { koaContext } = ctx;

  const { ip } = koaContext.request;
  const session = await SessionTokenModel.findOne({
    user: user._id,
    ip,
    channel,
    scope,
    isActive: true,
    isBlocked: false,
  }).lean<ISessionToken>();

  if (!session || isPast(session.expiresIn!)) return null;

  return session;
};

type SessionTokenArgs = {};

export const loadAll = async (context: GraphQLContext, args: SessionTokenArgs) => {
  const sessionTokens = SessionTokenModel.find().sort({ createdAt: -1 });

  return connectionFromMongoCursor({
    cursor: sessionTokens,
    context,
    args,
    loader: load,
  });
};

export const loadSessionTokensByIp = async (context: GraphQLContext, args: SessionTokenArgs, ip: string) => {
  const sessionTokens = SessionTokenModel.find({ ip }).sort({ createdAt: -1 });

  return connectionFromMongoCursor({
    cursor: sessionTokens,
    context,
    args,
    loader: load,
  });
};

export const loadSessionTokensByUser = async (context: GraphQLContext, args: SessionTokenArgs, user: string) => {
  const sessionTokens = SessionTokenModel.find({ user }).sort({ createdAt: -1 });

  return connectionFromMongoCursor({
    cursor: sessionTokens,
    context,
    args,
    loader: load,
  });
};
