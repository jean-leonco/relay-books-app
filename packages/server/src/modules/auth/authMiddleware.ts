import Application from 'koa';
import i18next from 'i18next';

import { GraphQLContext } from '../../types';

import { auth } from './sessionManagement';

const authMiddleware: Application.Middleware<any, GraphQLContext> = async (context, next) => {
  const { t } = context;

  const { abortRequest, unauthorized, user, error, status: resultStatus } = await auth(context);
  const status = resultStatus ? resultStatus : error ? 401 : 200;

  if (abortRequest || unauthorized) {
    context.status = status;

    context.body = {
      data: null,
      errors: [{ message: error || t('auth', 'InvalidSession'), severity: 'WARNING' }],
    };
    return;
  }

  context.user = user;

  if (user?.lang && i18next.isInitialized) await i18next.changeLanguage(user?.lang);

  return await next();
};

export default authMiddleware;
