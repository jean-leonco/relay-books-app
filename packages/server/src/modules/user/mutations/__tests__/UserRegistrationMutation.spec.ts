import { graphql } from 'graphql';

import { sanitizeTestObject } from '@booksapp/test-utils';

import { schema } from '../../../../graphql/schema';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  createUser,
  disconnectMongoose,
  getContext,
  gql,
} from '../../../../../test/helpers';
import { getPlatform, PLATFORM } from '../../../../common/utils';
import { SESSION_TOKEN_SCOPES } from '../../../sessionToken/SessionTokenModel';
import { generateToken } from '../../../auth/generateToken';
import { UserLoader } from '../../../../loader';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('UserRegistrationMutation', () => {
  it('should registrate an user', async () => {
    const name = 'jean';
    const email = 'jean@gmail.com';
    const password = '123456';

    const mutation = gql`
      mutation M($input: UserRegistrationInput!) {
        UserRegistration(input: $input) {
          token
          error
        }
      }
    `;

    const variables = {
      input: {
        name,
        email,
        password,
      },
    };
    const rootValue = {};
    const context = await getContext({ appplatform: PLATFORM.APP });
    const result = await graphql(schema, mutation, rootValue, context, variables);

    const user = await UserLoader.findUserByEmail(context, email);
    const token = await generateToken(context, user, getPlatform(context.appplatform), SESSION_TOKEN_SCOPES.SESSION);

    expect(result.errors).toBeUndefined();
    expect(result.data?.UserRegistration.error).toBe(null);
    expect(result.data?.UserRegistration.token).toBe(token);
    expect(user?.email.email).toBe(email);
    expect(user?.name).toBe(name);
    expect(sanitizeTestObject(result, ['token'])).toMatchSnapshot();
  });

  it('should not registrate an user with invalid email', async () => {
    const name = 'jean';
    const email = 'blablabla';
    const password = '123456';

    const mutation = gql`
      mutation M($input: UserRegistrationInput!) {
        UserRegistration(input: $input) {
          token
          error
        }
      }
    `;

    const variables = {
      input: {
        name,
        email,
        password,
      },
    };
    const rootValue = {};
    const context = await getContext({ appplatform: PLATFORM.APP });
    const result = await graphql(schema, mutation, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.UserRegistration.error).toBe('The email must be a valid email.');
    expect(result.data?.UserRegistration.token).toBe(null);
  });

  it('should not registrate an user using password with less than 6 characters', async () => {
    const name = 'jean';
    const email = 'blablabla';
    const password = '1234';

    const mutation = gql`
      mutation M($input: UserRegistrationInput!) {
        UserRegistration(input: $input) {
          token
          error
        }
      }
    `;

    const variables = {
      input: {
        name,
        email,
        password,
      },
    };
    const rootValue = {};
    const context = await getContext({ appplatform: PLATFORM.APP });
    const result = await graphql(schema, mutation, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.UserRegistration.error).toBe('Password must be at least 6 characters.');
    expect(result.data?.UserRegistration.token).toBe(null);
  });

  it('should not registrate user if email belongs to another user', async () => {
    const user = await createUser();
    const name = 'jean';
    const password = '123456';

    const mutation = gql`
      mutation M($input: UserRegistrationInput!) {
        UserRegistration(input: $input) {
          token
          error
        }
      }
    `;

    const variables = {
      input: {
        name,
        email: user.email.email,
        password,
      },
    };
    const rootValue = {};
    const context = await getContext({ appplatform: PLATFORM.APP });
    const result = await graphql(schema, mutation, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.UserRegistration.error).toBe('The email is already associated with another account.');
    expect(result.data?.UserRegistration.token).toBe(null);
  });
});
