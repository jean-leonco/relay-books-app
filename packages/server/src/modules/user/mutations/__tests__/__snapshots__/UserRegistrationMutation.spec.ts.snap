// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRegistrationMutation should registrate an user 1`] = `
Object {
  "data": Object {
    "UserRegistration": Object {
      "error": "EMPTY",
      "token": "FROZEN-TOKEN",
    },
  },
}
`;
